File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/network/JsonMessageGeneralSerializer.java
package com.github.dedis.popstellar.model.network.serializer.network;

import com.github.dedis.popstellar.model.network.method.message.MessageGeneral;
import com.github.dedis.popstellar.model.network.method.message.PublicKeySignaturePair;
import com.github.dedis.popstellar.model.network.method.message.data.Data;
import com.github.dedis.popstellar.model.network.serializer.JsonUtils;
import com.github.dedis.popstellar.model.objects.security.*;
import com.google.gson.*;
import com.google.gson.annotations.SerializedName;

import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.util.List;

public class JsonMessageGeneralSerializer
    implements JsonSerializer<MessageGeneral>, JsonDeserializer<MessageGeneral> {

  @Override
  public MessageGeneral deserialize(
      JsonElement json, Type typeOfT, JsonDeserializationContext context)
      throws JsonParseException {
    JsonMessageData jsonObject = context.deserialize(json, JsonMessageData.class);

    JsonElement dataElement =
        JsonParser.parseString(new String(jsonObject.data.getData(), StandardCharsets.UTF_8));
    Data data = context.deserialize(dataElement, Data.class);

    return new MessageGeneral(
        jsonObject.sender,
        jsonObject.data,
        data,
        jsonObject.signature,
        jsonObject.messageID,
        jsonObject.witnessSignatures);
  }

  @Override
  public JsonElement serialize(
      MessageGeneral src, Type typeOfSrc, JsonSerializationContext context) {
    JsonMessageData jsonObject =
        new JsonMessageData(
            src.getDataEncoded(),
            src.getSender(),
            src.getSignature(),
            src.getMessageId(),
            src.getWitnessSignatures());
    JsonElement result = context.serialize(jsonObject);
    JsonUtils.verifyJson(JsonUtils.GENERAL_MESSAGE_SCHEMA, result.toString());
    return result;
  }

  private static final class JsonMessageData {

    public final Base64URLData data;
    public final PublicKey sender;
    public final Signature signature;

    @SerializedName("message_id")
    public final MessageID messageID;

    @SerializedName("witness_signatures")
    public final List<PublicKeySignaturePair> witnessSignatures;

    private JsonMessageData(
        Base64URLData data,
        PublicKey sender,
        Signature signature,
        MessageID messageID,
        List<PublicKeySignaturePair> witnessSignatures) {
      this.data = data;
      this.sender = sender;
      this.signature = signature;
      this.messageID = messageID;
      this.witnessSignatures = witnessSignatures;
    }
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/network/JsonMessageGeneralSerializer.kt
package com.github.dedis.popstellar.model.network.serializer.network

import com.github.dedis.popstellar.model.network.method.message.MessageGeneral
import com.github.dedis.popstellar.model.network.method.message.PublicKeySignaturePair
import com.github.dedis.popstellar.model.network.method.message.data.Data
import com.github.dedis.popstellar.model.network.serializer.JsonUtils
import com.github.dedis.popstellar.model.network.serializer.JsonUtils.verifyJson
import com.github.dedis.popstellar.model.objects.security.Base64URLData
import com.github.dedis.popstellar.model.objects.security.MessageID
import com.github.dedis.popstellar.model.objects.security.PublicKey
import com.github.dedis.popstellar.model.objects.security.Signature
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonParser
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import com.google.gson.annotations.SerializedName
import java.lang.reflect.Type
import java.nio.charset.StandardCharsets

class JsonMessageGeneralSerializer :
  JsonSerializer<MessageGeneral>, JsonDeserializer<MessageGeneral> {
  @Throws(JsonParseException::class)
  override fun deserialize(
    json: JsonElement,
    typeOfT: Type,
    context: JsonDeserializationContext
  ): MessageGeneral {
    val jsonObject = context.deserialize<JsonMessageData>(json, JsonMessageData::class.java)
    val dataElement = JsonParser.parseString(jsonObject.data.data.toString(StandardCharsets.UTF_8))
    val data = context.deserialize<Data>(dataElement, Data::class.java)

    return MessageGeneral(
      jsonObject.sender,
      jsonObject.data,
      data,
      jsonObject.signature,
      jsonObject.messageID,
      jsonObject.witnessSignatures
    )
  }

  override fun serialize(
    src: MessageGeneral,
    typeOfSrc: Type,
    context: JsonSerializationContext
  ): JsonElement {
    val jsonObject =
      JsonMessageData(
        src.dataEncoded,
        src.sender,
        src.signature,
        src.messageId,
        src.witnessSignatures
      )
    val result = context.serialize(jsonObject)

    verifyJson(JsonUtils.GENERAL_MESSAGE_SCHEMA, result.toString())

    return result
  }

  private class JsonMessageData(
    val data: Base64URLData,
    val sender: PublicKey,
    val signature: Signature,
    @field:SerializedName("message_id") val messageID: MessageID,
    @field:SerializedName("witness_signatures") val witnessSignatures: List<PublicKeySignaturePair>
  )
}
